scalafmt-core/shared/src/main/scala/org/scalafmt/rewrite/RedundantBraces.scala

```diff
@@ -129,37 +129,49 @@ case object RedundantBraces extends Rewrite {
         implicit val builder = Seq.newBuilder[TokenPatch]
         builder += TokenPatch.Remove(lbrace)
         builder += TokenPatch.Remove(rbrace)
-        removeTrailingLF(lbrace.pos, rbrace)
-        ctx.tokenTraverser
-          .reverseFind(rbrace)(!Whitespace.unapply(_))
-          .foreach(t => removeTrailingLF(t.pos, rbrace))
+        removeLFToAvoidEmptyLine(rbrace)
         ctx.addPatchSet(builder.result(): _*)
       }
     case _ =>
   }
 
-  private def removeTrailingLF(
-      bodyEnd: Position,
-      close: Token
-  )(implicit builder: PatchBuilder, ctx: RewriteCtx): Unit =
-    if (close.pos.startLine != bodyEnd.endLine) {
-      import ctx.tokenTraverser._
-      val next = nextToken(close)
-      if (next.is[LF])
-        builder += TokenPatch.Remove(next)
+  private def isLFSkipWhitespace(token: Token): Option[Boolean] =
+    token match {
+      case _: LF => Some(true)
+      case Whitespace() => None
+      case _ => Some(false)
     }
 
-  private def processBlock(
-      b: Term.Block
-  )(implicit ctx: RewriteCtx): Unit =
-    if (b.tokens.nonEmpty) {
-      val open = b.tokens.head
-      if (open.is[LeftBrace]) {
-        val close = b.tokens.last
-        if (removeBlock(b) && close.is[RightBrace]) {
-          val endPos = if (b.stats.isEmpty) b.pos else b.stats.last.pos
-          implicit val builder = Seq.newBuilder[TokenPatch]
-          removeTrailingLF(endPos, close)
+  private def removeLFToAvoidEmptyLine(
+      token: Token
+  )(implicit builder: PatchBuilder, ctx: RewriteCtx): Unit = {
+    if (ctx.tokenTraverser
+        .findBefore(token)(isLFSkipWhitespace)
+        .isDefined)
+      ctx.tokenTraverser
+        .findAfter(token)(isLFSkipWhitespace)
+        .foreach(builder += TokenPatch.Remove(_))
+  }
+
+  private def processBlock(b: Term.Block)(implicit ctx: RewriteCtx): Unit =
+    b.tokens.headOption.filter(_.is[LeftBrace]).foreach { open =>
+      val close = b.tokens.last
+      if (close.is[RightBrace] && removeBlock(b)) {
+        implicit val builder = Seq.newBuilder[TokenPatch]
+        val ok =
+          if (b.parent.exists(_.is[Term.ApplyInfix])) {
+            /* for infix, we will preserve the block unless the closing brace
+             * follows a non-whitespace character on the same line as we don't
+             * break lines around infix expressions.
+             * we shouldn't join with the previous line (which might also end
+             * in a comment), and if we keep the break before the right brace
+             * we are removing, that will likely invalidate the expression. */
+            ctx.tokenTraverser.findBefore(close)(isLFSkipWhitespace).isEmpty
+          } else {
+            removeLFToAvoidEmptyLine(close)
+            true
+          }
```

Comment:
So this tries to fix.

```
  a ~> {
    b
  } ~> c
is rewritten as

  a ~>
    b
  ~> c
instead of

  a ~>
    b ~> c
```

and


```
Infix lambda calls:
  a map { _.toString }
to

  a map _.toString
```

from https://github.com/scalameta/scalafmt/issues/1633, and https://github.com/scalameta/scalafmt/issues/1631, right?

---

I think this leads to invalid formatting result: for example

if we format the following code:

```scala
val x = xs map { _.toString }
```

to

```scala
val x = xs map _.toString
```

which doesn't compile.

Maybe we should stop remove braces for `Term.ApplyInfix` like `Term.Apply` doesn't https://github.com/adampauls/scalafmt/blob/7a3ace0a8209008ad13916e19a0919e425ed43f0/scalafmt-core/shared/src/main/scala/org/scalafmt/rewrite/RedundantBraces.scala#L201-L205 what do you think about this?

Reply:
i was trying to remove as little functionality as possible. i am leaning towards what looks like a reasonably safe option of basically allowing only another infix with a parent infix, using the precedence logic that was there before. thoughts?